### ISSUE TITLE
W1b: `reproject_disparity_to_3D` — `u` and `v` are swapped, and the existing real-data pin encodes the swap

### ISSUE THREAD

**Issue by ducha-aiki** [kornia/kornia#4269]
`StereoCamera.reproject_disparity_to_3D` and the module-level `reproject_disparity_to_3D` (they share one body) unbind the pixel meshgrid as `v, u = torch.unbind(uv, dim=-1)`. `create_meshgrid(..., normalized_coordinates=False)` returns `(x, y)`, so `uv[..., 0]` is the **column** and `uv[..., 1]` is the **row**. The code therefore feeds the column index into `v` and the row index into `u`, and the returned `X` is computed from the row while `Y` is computed from the column — the opposite of `cv2.reprojectImageTo3D`, whose semantics this function was added to provide (#2042, closed 2025-12-29).

Every image gets wrong 3D coordinates, silently. What comes out is the OpenCV answer with the two pixel indices **transposed**: kornia's value at `(row, col)` is what `cv2.reprojectImageTo3D` puts at `(col, row)` — exactly `X = (row - cx) Z / fx`, `Y = (col - cy) Z / fy` with `Z = fx · tx / d`. `Q`'s own focal swap (row 0 carries `fy·b` and row 1 `fx·b`) compensates for the homogeneous divide, so `X` is still scaled by `1/fx` and `Y` by `1/fy`; what moved is only which *index* feeds which coordinate. A symmetric camera does **not** make it cancel: on a square image with `fx == fy` and `cx == cy`, transposing the index pair is a no-op only where `row == col`, so the diagonal agrees with OpenCV and nothing else does — 0.2 apart on the 5 × 5 rig in the repro below.

Three independent reproductions below: a synthetic asymmetric grid against the hand-computed OpenCV formula; the axis-dependence check (`X` must vary with the column and `Y` with the row, and they do the reverse); and **the repo's own real-data test fixture**, whose stored ground truth is the swapped value — `tests/geometry/camera/test_stereo.py::_RealTestData` stores a `(1, 10, 1, 1)` disparity (10 rows, 1 column, though its comment says "first 10 cols of 1 row"), and `test_reproject_disparity_to_3D_real` passes *because of* the bug.

## Repro

```python
import sys, os, torch
from kornia.geometry.camera import StereoCamera
from kornia.utils import create_meshgrid

def stereo(fx=100.0, fy=100.0, cx=4.0, cy=3.0, tx=0.5):
    KL = torch.zeros(1, 3, 4); KL[:, 0, 0] = fx; KL[:, 1, 1] = fy
    KL[:, 0, 2] = cx; KL[:, 1, 2] = cy; KL[:, 2, 2] = 1.0
    KR = KL.clone(); KR[:, 0, 3] = -tx * fx
    return StereoCamera(KL, KR)

def opencv_hand(row, col, d=10.0, fx=100.0, fy=100.0, cx=4.0, cy=3.0, Tb=0.5):
    """cv2.reprojectImageTo3D semantics: u is the COLUMN, v is the ROW."""
    Z = fx * Tb / d
    return [round((col - cx) * Z / fx, 6), round((row - cy) * Z / fy, 6), round(Z, 6)]

# H = 3 rows, W = 5 cols; cx = 4 != cy = 3, so the swap cannot cancel
s = stereo()
pts = s.reproject_disparity_to_3D(torch.full((1, 3, 5, 1), 10.0))
for (r, c) in ((0, 2), (1, 0), (2, 4)):
    print(f"(row {r}, col {c})  kornia {[round(v,6) for v in pts[0, r, c].tolist()]}   opencv {opencv_hand(r, c)}")
print("X along a row   (col varies, row fixed):", [round(v, 6) for v in pts[0, 0, :, 0].tolist()])
print("X down a column (row varies, col fixed):", [round(v, 6) for v in pts[0, :, 0, 0].tolist()])
print("Y along a row   (col varies, row fixed):", [round(v, 6) for v in pts[0, 0, :, 1].tolist()])
print("create_meshgrid(...)[..., 0] is the COLUMN index:", create_meshgrid(3, 5, normalized_coordinates=False)[0, :, :, 0].tolist())

# the whole map is the OpenCV map with the two indices transposed -- shown on fx = 100 != fy = 50
asym = stereo(fy=50.0)
pts_a = asym.reproject_disparity_to_3D(torch.full((1, 3, 5, 1), 10.0))
rows = torch.arange(3.0).view(3, 1).expand(3, 5)
cols = torch.arange(5.0).view(1, 5).expand(3, 5)
transposed = torch.stack([(rows - 4.0) * 5.0 / 100.0, (cols - 3.0) * 5.0 / 50.0, torch.full((3, 5), 5.0)], -1)[None]
print("fx != fy: max |kornia - opencv_at(u=row, v=col)|:", round((pts_a - transposed).abs().max().item(), 9))

# a symmetric camera on a square image does not make it cancel
sq = stereo(cx=3.0, cy=3.0)
pts_s = sq.reproject_disparity_to_3D(torch.full((1, 5, 5, 1), 10.0))
r5 = torch.arange(5.0).view(5, 1).expand(5, 5)
c5 = torch.arange(5.0).view(1, 5).expand(5, 5)
opencv5 = torch.stack([(c5 - 3.0) * 0.05, (r5 - 3.0) * 0.05, torch.full((5, 5), 5.0)], -1)[None]
print("fx == fy, cx == cy, square: max |kornia - opencv|:", round((pts_s - opencv5).abs().max().item(), 9))
print("  the same on the diagonal pixel (2, 2)          :", round((pts_s[0, 2, 2] - opencv5[0, 2, 2]).abs().max().item(), 9))

# The repo's own "real data" pin encodes the swap
sys.path.insert(0, os.getcwd())
from tests.geometry.camera.test_stereo import _RealTestData
dev, dt = torch.device("cpu"), torch.float64
left = _RealTestData._get_real_left_camera(1, dev, dt)
right = _RealTestData._get_real_right_camera(1, dev, dt)
disp = _RealTestData._get_real_disparity(1, dev, dt)
gt = _RealTestData._get_real_point_cloud(1, dev, dt)
out = StereoCamera(left, right).reproject_disparity_to_3D(disp)
fx, fy, cx, cy = left[0, 0, 0].item(), left[0, 1, 1].item(), left[0, 0, 2].item(), left[0, 1, 2].item()
Tb = -right[0, 0, 3].item() / fx
print("stored disparity shape (B,H,W,1)      :", tuple(disp.shape), "-> 10 rows, 1 column")
for r in (0, 1, 2):
    d = disp[0, r, 0, 0].item()
    print(f"  row {r}: kornia {[round(v,4) for v in out[0, r, 0].tolist()]}"
          f"  stored GT {[round(v,4) for v in gt[0, r, 0].tolist()]}"
          f"  opencv {[round(v,4) for v in opencv_hand(r, 0, d, fx, fy, cx, cy, Tb)]}")
```

## Observed

```
(row 0, col 2)  kornia [-0.2, -0.05, 5.0]   opencv [-0.1, -0.15, 5.0]
(row 1, col 0)  kornia [-0.15, -0.15, 5.0]   opencv [-0.2, -0.1, 5.0]
(row 2, col 4)  kornia [-0.1, 0.05, 5.0]   opencv [0.0, -0.05, 5.0]
X along a row   (col varies, row fixed): [-0.2, -0.2, -0.2, -0.2, -0.2]
X down a column (row varies, col fixed): [-0.2, -0.15, -0.1]
Y along a row   (col varies, row fixed): [-0.15, -0.1, -0.05, -0.0, 0.05]
create_meshgrid(...)[..., 0] is the COLUMN index: [[0.0, 1.0, 2.0, 3.0, 4.0], [0.0, 1.0, 2.0, 3.0, 4.0], [0.0, 1.0, 2.0, 3.0, 4.0]]
fx != fy: max |kornia - opencv_at(u=row, v=col)|: 0.0
fx == fy, cx == cy, square: max |kornia - opencv|: 0.200000003
  the same on the diagonal pixel (2, 2)          : 0.0
stored disparity shape (B,H,W,1)      : (1, 10, 1, 1) -> 10 rows, 1 column
  row 0: kornia [-30.2769, -19.3972, 80.4424]  stored GT [-30.2769, -19.3972, 80.4424]  opencv [-30.2769, -19.3972, 80.4424]
  row 1: kornia [-30.1945, -19.3961, 80.4377]  stored GT [-30.1945, -19.3961, 80.4377]  opencv [-30.2752, -19.3154, 80.4377]
  row 2: kornia [-30.112, -19.395, 80.4331]  stored GT [-30.112, -19.395, 80.433]  opencv [-30.2734, -19.2335, 80.4331]
```

## Expected

`u, v = torch.unbind(uv, dim=-1)` — `u` the column, `v` the row — matching `cv2.reprojectImageTo3D` and every other consumer of `create_meshgrid` in kornia. That is a one-line focused fix and it is welcome as a PR, with two things that must land in the same change:

1. **`_RealTestData`'s stored point cloud must be re-derived.** It currently encodes the swapped answer (rows 1 and 2 above), so `test_reproject_disparity_to_3D_real` will fail after the fix. The correct values are the `opencv` column: for row 1, `[-30.2752, -19.3154, 80.4377]` rather than the stored `[-30.1945, -19.3961, 80.4377]`. Row 0 is degenerate (row = col = 0) and is unchanged, which is why the fixture looked healthy.
2. **`Q` must be left alone.** Its row 0 carries `fy·tx` and row 1 `fx·tx`, and the transposed-map check above settles what that is: a compensation for the homogeneous divide (by `-fy · disparity`), not a second half of the same swap — with the indices transposed, `X` still comes out scaled by `1/fx` and `Y` by `1/fy`. So the `unbind` change alone is the whole fix; swapping `Q`'s rows as well would reintroduce the error. Confirm that in the same pass rather than assuming it.

Batch 5 leaves the existing assertion untouched and adds, beside it, a `test_wart_*` pinning today's swapped output and a strict `xfail` `test_convention_*` asserting the hand-computed OpenCV value, so the fix flips the pair.

## Classification

**Broken output** — wrong 3D coordinates on every non-square or off-centre image, with no error. Idiom: 7 (axis-order slip hidden by symmetric test data) — ratified 2026-09-10; the earlier instance is #4058.

## Disposition

n/a — single symbol. The method is a one-line forwarder to the free function, so both are fixed by the same change.

## Environment

kornia 0.9.0rc1 @ `92e5fe8d` (main), torch 2.14.0, python 3.11, cpu. Found during the camera conventions audit (batch 5, tracking issue #4262); the batch documents the behavior below as it is and pins it, and does not change it.

Posted on behalf of @ducha-aiki by Claude (Opus 5).

**Comment by ducha-aiki** [comment]
Correction to item 1 of the Expected section ("`_RealTestData`'s stored point cloud must be re-derived"): the stored literals are **correct OpenCV output and should stay**. Only the fixture's layout is transposed.

Executed on `main` `8f292308` (torch 2.14.0, cpu, float64) with the fixture as stored, disparity `(1, 10, 1, 1)` and point cloud `(1, 10, 1, 3)`:

| computation | max abs gap to the stored point cloud |
|---|---|
| current code (`v, u` unbind) on the stored layout | 8.77e-5 |
| corrected `u, v` unbind on the stored layout | 0.726 |
| corrected `u, v` unbind with disparity re-laid as `(1, 1, 10, 1)` and the cloud as `(1, 1, 10, 3)` | 8.77e-5 |
| hand `cv2.reprojectImageTo3D` formula for a one-row, ten-column strip (`u = 0..9`, `v = 0`) | 8.77e-5 |

So the fixture's own comment, "First 10 cols of 1 row", is right about the data: the literals are what OpenCV returns for a one-row strip, and the `.permute(0, 2, 3, 1)` at the end of `_get_real_disparity` is what turns that row into ten rows of one column, which the index swap then cancels. The fix is to correct the layout (drop that permute, lay the point cloud out the same way) and keep every number; regenerating the literals from kornia's output on the transposed layout, which is what item 1 asked for, would replace a real OpenCV check with a self-pin. #4366 already does the layout fix. The #4317 warning on `StereoCamera` and the strict-xfail comment beside `test_convention_reproject_disparity_uses_the_column_as_u_4269` said "the stored ground truth encodes the swap"; both are reworded in #4317 to say this.

Posted on behalf of @ducha-aiki by Claude (Claude Fable 5.1).

**Comment by ducha-aiki** [comment]
Re-tagged: the classification now reads **idiom 7** (axis-order slip hidden by symmetric test data), ratified 2026-09-10 as the seventh recurring root cause in the conventions programme; #4058 is the earlier instance. The detector that comes with it: every axis-order or field-order claim is executed on an asymmetric fixture and, where it is a labelling, checked under relabelling.

Posted on behalf of @ducha-aiki by Claude (Fable 5.1).
